harness/determined/_generic.py

```diff
@@ -0,0 +1,259 @@
+import contextlib
+import datetime
+import logging
+import math
+import pathlib
+from typing import Any, Dict, Iterator, cast
+
+import determined as det
+from determined import tensorboard, workload
+from determined.common import check, constants, storage
+
+
+class _GenericContext:
+    """
+    _GenericContext will someday evolve into a core part of the Generic API.
+
+    Today, _GenericContext is a storage place for objects which will probably be needed for the
+    Generic API but which are also used in the old layers architecture.
+    """
+
+    def __init__(
+        self,
+        env: det.EnvContext,
+        dist: det.DistributedContext,
+    ) -> None:
+        self._env = env
+        self._dist = dist
+
+        self._storage_mgr = storage.build(
+            env.experiment_config["checkpoint_storage"],
+            container_path=None if not env.on_cluster else constants.SHARED_FS_CONTAINER_PATH,
+        )
+
+        self._tensorboard_mgr = tensorboard.build(
+            env.det_cluster_id,
+            env.det_experiment_id,
+            env.det_trial_id,
+            env.experiment_config["checkpoint_storage"],
+            container_path=None if not env.on_cluster else constants.SHARED_FS_CONTAINER_PATH,
+        )
+
+        self._tensorboard_writer = tensorboard.get_metric_writer()
+
+    @contextlib.contextmanager
+    def _download_initial_checkpoint(self, checkpoint: Dict) -> Iterator[pathlib.Path]:
+        """
+        Wrap a storage_mgr.restore_path() context manager, but only download/cleanup on the
+        local chief.
+        """
+
+        metadata = storage.StorageMetadata.from_json(checkpoint)
+        logging.info("Restoring trial from checkpoint {}".format(metadata.storage_id))
+
+        restore_path = self._dist._local_chief_contextmanager(self._storage_mgr.restore_path)
+        with restore_path(metadata) as path:
+            yield path
+
+    @staticmethod
+    def _current_timestamp() -> datetime.datetime:
+        return datetime.datetime.now(datetime.timezone.utc)
+
+    def _after_training(
+        self,
+        wkld: workload.Workload,
+        start_time: datetime.datetime,
+        response: workload.Response,
+    ) -> workload.Response:
+        if self._dist.get_rank() != 0:
+            return response
+
+        check.is_not_instance(response, workload.Skipped, "Chief skipped a workload.")
+
+        response = cast(workload.Metrics, response)
+        metrics = response["metrics"]
+        metrics = cast(workload.Metrics, metrics)
+
+        if response.get("invalid_hp", False):
+            out_response = {
+                "type": "WORKLOAD_COMPLETED",
+                "workload": wkld,
+                "start_time": start_time,
+                "end_time": self._current_timestamp(),
+                "metrics": metrics,
+                "exited_reason": "INVALID_HP",
```

Comment:
Ok, so the particular commit you are reviewing here was written with the specific goal of taking the contents of the WorkloadManager, which by its nature lived in-between two layers of workload message duplication, and moving it to live inside the worker processes, after all duplication (see [sprinkle rfc architecture diagram](https://github.com/rb-determined-ai/determined/blob/sprinkle-spec/sprinkle/arch.md)).  The focus was preserving exact functionality in a new architecture, not making improvements to some old (sometimes questionable) code.

The initial commit on this PR could actually land on master as a standalone commit, but it seemed like an unnecessarily high risk commit, so I left it in the feature branch.

So initially I was going to say "please don't make me change this because it's just backwards compatibility code with some old behavior that is going to be deprecated soon".  But now that I'm further along with push architecture, I think each of the `_after_training()`, `_after_validation()`, and `_after_checkpoint()` actually need to be broken up and their contents moved into a few different places for the Generic API, so it may as well happen now.

For example, the validation metrics filtering should go in `_training.Training.report_validation_metrics()`, since direct consumers of the Generic API should benefit from the same logic.  But the check for the presence of the searcher metric itself should go in the `WorkloadSequencer`, since only the Trial APIs take the searcher metric implicitly from validation metrics; direct consumers of the Generic API would be reporting the searcher metric explicitly.

Anyway, I'm working on that change now.  So expect a bunch of change to `_generic.py`.  Probably won't be ready today though.

Reply:
Makes sense.

QQ:
>direct consumers of the Generic API would be reporting the searcher metric explicitly.

I assumed it'd look like: 
```
context.training.report_validation_metrics(
  batches_trained=batches_trained,
  metrics={"loss": eval_loss, "metric1": metric1, "metric2": metric2},
)
```

Then, checking for the searcher metric presence still seems useful in `_training.Training.report_validation_metrics()` because it may prevent user's programming error.

Reply:
Actually, it's pretty darn hard for the master to 'steal' the searcher metric from reported validation metrics.  For example, how many `total_batches` corresponds to the comletion of the first epoch?  That's impossible to answer in general.  So instead the completion of a searcher operation means that you have explicitly report the searcher metric (a single scalar) via searcher-specific API.

That also moves us in the direction of not requiring a searcher metric to be named in the experiment config explicitly at all.  We talked about moving from a required searcher metric to having a "preferred sorting metric" or something, which in the generic API could be left undefined, in which case when you asked for e.g. the "top n checkpoints" you would just have to provide the sorting criteria at request time.

Reply:
Okay. So:
```
context.training.report_validation_metrics(
  metrics={"loss": eval_loss, "metric1": metric1, "metric2": metric2},
)
```

(I liked an option to switch the searcher metric without changing the model code and only the experiment config: experiment reports a number of metrics in bulk, a few of them are similarly "important" to me, I experiment with adaptive by switching  the searcher metric.)

Reply:
Yes, but realistically too much of our system is built on having a known x axis for us to drop the `total_batches` argument.  Some day it would be nice to allow the x axis to be logged as a normal "metric", and you could replot with arbitrary x axes, but we're not so close to that world just yet.

Re configurable searcher metric: just set it in the config file and write your python code to choose one metric or another metric based on the config file.